### ISSUE TITLE
JMX agent discovery does not work on Java 8

### ISSUE THREAD

**Issue by electrum** [airlift/airlift#309]
2015-05-10T19:04:59.006-0700	WARN	main	io.airlift.jmx.JmxAgent	Cannot determine if JMX agent is already running (not an Oracle JVM?). Will try to start it manually.

**Comment by damiencarol** [comment]
Any news on this one?

**Comment by martint** [comment]
This is actually not a bug. If the VM is not started with the `com.sun.management.jmxremote` options the jmx agent built into the VM won't be running. Unfortunately, there's no reliable way to distinguish between the agent not being started and the current version of the VM using a different mechanism to initialize it. Hence, the best we can do is report a message describing the situation.

We can probably improve the message, and I have a couple of ideas on how to improve the detection heuristics, but this is working as expected.

**Comment by damiencarol** [comment]
@martint thanks for the details. I think we missed the `com.sun.management.jmxremote` option in Java 1.8 migration.

**Comment by electrum** [comment]
We do need to fix the message and ideally update the detection. Maybe just always try to start it if configured to do so, otherwise do nothing? I actually think the default behavior of always starting with no way to disable is wrong.

**Comment by martint** [comment]
> We do need to fix the message and ideally update the detection. 

Agreed

> Maybe just always try to start it if configured to do so, otherwise do nothing? 

Well, given that "if configure to do so" is always true, that's exactly what we do :)

> I actually think the default behavior of always starting with no way to disable is wrong.

I don't know about that. Ages ago we decided that all airlift-based services than include JmxModule have jmx. Turning that off was never meant to be a runtime option (once you do that, you lose the ability to do monitoring)

**Comment by electrum** [comment]
I agree that having it enabled by default is good. There is the problem that JVM opens a third random mystery port, so users may want to turn of the JMX RMI agent (perhaps for security reasons).

Is the RMI connector actually needed if you have jstatd?

Related post: https://groups.google.com/d/msg/presto-users/UWxNT89g160/Wz9yKnxyS1wJ
